google/cloud/aiplatform/experimental/vertex_model/serializers/dataloaders.py

```diff
@@ -15,35 +15,256 @@
+       a run-time bucket
+
+    Args:
+        artifact_uri (str): the GCS bucket where the serialized object will reside.
+        dataloader_path (str): the path where the origin data used to construct the DataLoader
+                               resides.
+        obj (torch.utils.data.DataLoader): the pytorch DataLoader to serialize.
+        dataset_type (str): the intended use of the dataset (ie. training, testing)
+
+    Returns:
+        The GCS path pointing to the serialized DataLoader, the GCS path pointing to the
+        serialized origin data.
+    """
+
+    if not artifact_uri.startswith("gs://"):
+        local_path = artifact_uri + "my_" + dataset_type + "_dataloader.pth"
+        torch.save(obj, local_path)
+        return local_path, dataloader_path
+
+    dataloader_path = pathlib.Path(dataloader_path)
+
+    gcs_bucket, gcs_blob_prefix = utils.extract_bucket_and_prefix_from_gcs_path(
+        artifact_uri
+    )
+
+    data_blob_path = dataset_type + "_" + dataloader_path.name
+
+    if gcs_blob_prefix:
+        data_blob_path = "/".join([gcs_blob_prefix, data_blob_path])
+
+    client = storage.Client(
+        project=initializer.global_config.project,
+        credentials=initializer.global_config.credentials,
+    )
+
+    bucket = client.bucket(gcs_bucket)
+    data_blob = bucket.blob(data_blob_path)
+    data_blob.upload_from_filename(str(dataloader_path))
+
+    data_gcs_path = "".join(
+        ["gs://", "/".join([data_blob.bucket.name, data_blob.name])]
+    )
+
+    path = serialize_to_tmp_and_copy_to_gcs(
+        "my_" + dataset_type + "_dataloader.pth",
+        gcs_bucket,
+        gcs_blob_prefix,
+        torch.save,
+    )
+
+    return path, data_gcs_path
+
+
+def _serialize_dataloader(
+    artifact_uri: str, obj: torch.utils.data.DataLoader, dataset_type: str
+) -> (str, str):
+    """Serializes DataLoader object to GCS and stores remotely-sourced data in
+       a run-time bucket. Determines which helper method to use by introspecting
+       the user-specified DataLoader object.
+
+    Args:
+        artifact_uri (str): the GCS bucket where the serialized object will reside.
+        obj (torch.utils.data.DataLoader): the DataLoader to be serialized
+        dataset_type (str): the intended use of the dataset (ie. training, testing)
+
+    Returns:
+        The GCS path pointing to the serialized DataLoader, the GCS path pointing to the
+        serialized origin data.
+    """
+    my_dataset = obj.dataset
+    root = getattr(my_dataset, "root", None)
+
+    if root is None:
+        return _serialize_local_dataloader(artifact_uri, root, obj, dataset_type)
+
+    # Decide whether to pass to remote or local serialization
+    if root.startswith("gs://"):
+        return _serialize_remote_dataloader(artifact_uri, root, obj, dataset_type)
+    else:
+        return _serialize_local_dataloader(artifact_uri, root, obj, dataset_type)
+
+
+def _deserialize_dataloader(artifact_uri: str) -> DataLoader:
```

Comment:
How is the case where the local dataset is also serialized to GCS handled (`data_gcs_path` in the serializer)?

It seems like it needs to be copied from GCS and possibly have the local reference in the object replaced.

Right now it seems like we are only deserializing `path`, the torch object, from the serializer.